### PR TITLE
revert some code for #4726 / remove unnecessary coercion in physical plans

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

from the https://github.com/apache/arrow-datafusion/pull/4726#issuecomment-1365869414

If we want to compare the timestamp with date, we just need to add logic in `type_coercion`, and don't need to do type coercion in the physical phase.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/type_coercion/binary.rs [hunk 1]
```diff
@@ -575,6 +575,7 @@ fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataTyp
         },
         (Timestamp(_, tz), Utf8) => Some(Timestamp(TimeUnit::Nanosecond, tz.clone())),
         (Utf8, Timestamp(_, tz)) => Some(Timestamp(TimeUnit::Nanosecond, tz.clone())),
+        // TODO: need to investigate the result type for the comparison between timestamp and date
```
Comment (liukun4515) [inline datafusion/expr/src/type_coercion/binary.rs:578]: https://github.com/apache/arrow-datafusion/issues/4644#issuecomment-1365869161
we need to get the behavior for the comparison in the PG.

**Comment by liukun4515** [comment]
cc @alamb @comphead

File: datafusion/expr/src/type_coercion/functions.rs [hunk 2]
```diff
@@ -177,35 +177,12 @@ pub fn can_coerce_from(type_into: &DataType, type_from: &DataType) -> bool {
         Timestamp(TimeUnit::Nanosecond, None) => {
             matches!(type_from, Null | Timestamp(_, None))
         }
-        Date32 => {
```
Comment (alamb) [inline datafusion/expr/src/type_coercion/functions.rs:180]: why this change?

Though I admit that the fact that all tests still pass is concerning 🤔

**Review by alamb — APPROVED** [review]
👍 I like it 

(I made a similar version here https://github.com/apache/arrow-datafusion/pull/4745 before I found this PR)

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (comphead) [reply datafusion/expr/src/type_coercion/binary.rs:578]: @liukun4515  comparison is usually boolean type, did you mean what is the common datatype for casting between timestamp and date in diff scenarios?

File: datafusion/expr/src/type_coercion/functions.rs [hunk 3]
```diff
@@ -177,35 +177,12 @@ pub fn can_coerce_from(type_into: &DataType, type_from: &DataType) -> bool {
...
         Utf8 | LargeUtf8 => true,
         Null => can_cast_types(type_from, type_into),
         _ => false,
     }
 }
 
-/// Returns a common coerced datatype between 2 given datatypes
-///
-/// See the module level documentation for more detail on coercion.
-pub fn get_common_coerced_type(
```
Comment (comphead) [inline datafusion/expr/src/type_coercion/functions.rs:192]: I think this method can be of some interest, if the user wants to check the common coerced type between 2 datatypes?

Comment (alamb) [reply datafusion/expr/src/type_coercion/functions.rs:192]: It may be worth waiting for a specific use before leaving it in 🤔

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:578]: Yes
the result of comparison is bool type, but if left and right are the same data type, we should get the common type for left and right through the `type coercion`.
For example, INT32>INT64, the common type is INT64, and we should cast the left to INT64.

Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:578]: The type of `type coercion` is decided by `input types` and `op`.

File: datafusion/expr/src/type_coercion/functions.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/expr/src/type_coercion/functions.rs:180]: This was added from the commit https://github.com/comphead/arrow-datafusion/blob/4761802a4bc46d377406aa20767c4af66b80d65b/datafusion/expr/src/type_coercion/functions.rs#L180 in the https://github.com/apache/arrow-datafusion/pull/4726

I just revert useless codes adding from the PR https://github.com/apache/arrow-datafusion/pull/4726

File: datafusion/expr/src/type_coercion/functions.rs (hunk 3 above)
Comment (liukun4515) [reply datafusion/expr/src/type_coercion/functions.rs:192]: > I think this method can be of some interest, if the user wants to check the common coerced type between 2 datatypes?

the common coerced type will be decided by `input exprs/input types` and `op`, if the `op` is different and the same `input types/input exprs` may get the different coerced type.

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:578]: > @liukun4515 comparison is usually boolean type, did you mean what is the common datatype for casting between timestamp and date in diff scenarios?

I think it's not about `casting` from one type to other type, it is related the `type inference` for different input types and different operation or functions.

Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:578]: I try the comparison timestamp with date in the Spark, and find the coerced type is timestamp.

```
spark-sql> explain extended select now()>cast('2000-01-01' as date);
== Parsed Logical Plan ==
'Project [unresolvedalias(('now() > cast(2000-01-01 as date)), None)]
+- OneRowRelation

== Analyzed Logical Plan ==
(now() > CAST(2000-01-01 AS DATE)): boolean
Project [(now() > cast(cast(2000-01-01 as date) as timestamp)) AS (now() > CAST(2000-01-01 AS DATE))#49]
+- OneRowRelation

== Optimized Logical Plan ==
Project [true AS (now() > CAST(2000-01-01 AS DATE))#49]
+- OneRowRelation

== Physical Plan ==
*(1) Project [true AS (now() > CAST(2000-01-01 AS DATE))#49]
+- *(1) Scan OneRowRelation[]
```


cc @comphead @alamb 

But current implementation of coerced type is date, we can fix this in the follow up PR.

Comment (liukun4515) [reply datafusion/expr/src/type_coercion/binary.rs:578]: https://github.com/apache/arrow-datafusion/issues/4761

File: datafusion/expr/src/type_coercion/functions.rs (hunk 3 above)
Comment (alamb) [reply datafusion/expr/src/type_coercion/functions.rs:192]: Let's merge this PR then -- we can always recover `get_common_coerced_type` from this PR or git history if we need it.

File: datafusion/expr/src/type_coercion/binary.rs (hunk 1 above)
Comment (alamb) [reply datafusion/expr/src/type_coercion/binary.rs:578]: Timestamp is probably consistent with other coercions that ensure no data is truncated (so going Date --> Timestamp is good as the the Timestamp could  be losslessly converted back to a Date). However, going from Timestamp --> Date would truncate the precision of the original timesteamp,
